Archipelago.MultiClient.Net/Cache/DataPackageFileSystemCache.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Archipelago.MultiClient.Net.Enums;
+using Archipelago.MultiClient.Net.Exceptions;
+using Archipelago.MultiClient.Net.Helpers;
+using Archipelago.MultiClient.Net.Models;
+using Archipelago.MultiClient.Net.Packets;
+using Newtonsoft.Json;
+
+namespace Archipelago.MultiClient.Net.Cache
+{
+    internal class DataPackageFileSystemCache : IDataPackageCache
+    {
+        private const string DataPackageFileName = "datapackagecache.archipelagocache";
+        private readonly ArchipelagoSocketHelper socket;
+        private readonly string CacheFolder = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+        private RoomInfoPacket roomInfoPacket;
+        private DataPackage dataPackage;
+
+        private readonly object fileAccessLockObject = new object();
+
+        public DataPackageFileSystemCache(ArchipelagoSocketHelper socket)
+        {
+            this.socket = socket;
+
+            socket.PacketReceived += Socket_PacketReceived;
+        }
+
+        private void Socket_PacketReceived(ArchipelagoPacketBase packet)
+        {
+            if (packet.PacketType == ArchipelagoPacketType.RoomInfo)
+            {
+                roomInfoPacket = (RoomInfoPacket)packet;
+                var invalidated = GetCacheInvalidatedGames(roomInfoPacket);
+
+                if (invalidated.Any())
+                {
+                    var exclusions = roomInfoPacket.DataPackageVersions.Select(x => x.Key).Except(invalidated);
+
+                    socket.SendPacket(new GetDataPackagePacket()
+                    {
+                        Exclusions = exclusions.ToList()
+                    });
+                }
+            }
+            else if (packet.PacketType == ArchipelagoPacketType.DataPackage)
+            {
+                var packagePacket = (DataPackagePacket)packet;
+                SaveDataPackageToCache(packagePacket.DataPackage);
+            }
+        }
+
+        public bool TryGetDataPackageFromCache(out DataPackage package)
+        {
+            if (dataPackage != null)
+            {
+                package = dataPackage;
+                return true;
+            }
+            lock (fileAccessLockObject)
+            {
+                var dataPackagePath = Path.Combine(CacheFolder, DataPackageFileName);
+                if (File.Exists(dataPackagePath))
+                {
+                    try
+                    {
+                        string fileText = File.ReadAllText(dataPackagePath);
+                        package = JsonConvert.DeserializeObject<DataPackage>(fileText);
+                        dataPackage = package;
+                        return true;
+                    }
+                    catch (Exception e)
+                    {
+                        throw new CacheLoadFailureException("Could not load data package cache from file system.", e);

```

Comment:
maybe not throw this here, but instead threat it as it no cache is available that saves a lot of places that otherwise have to try/catch protect themzelf for this exception

Reply:
Exception was removed and threated as if there is no cache available